packages/vite/src/node/plugins/define.ts

```diff
@@ -145,36 +135,88 @@ export function definePlugin(config: ResolvedConfig): Plugin {
         return
       }
 
-      const [replacements, pattern] = ssr ? ssrPattern : defaultPattern
+      const [define, pattern] = ssr ? ssrPattern : defaultPattern
+      if (!pattern) return
 
-      if (!pattern) {
-        return null
-      }
+      // Check if our code needs any replacements before running esbuild
+      pattern.lastIndex = 0
+      if (!pattern.test(code)) return
 
-      if (ssr && !isBuild) {
-        // ssr + dev, simple replace
-        return code.replace(pattern, (_, match) => {
-          return '' + replacements[match]
-        })
-      }
+      return await replaceDefine(code, id, define, config)
+    },
+  }
+}
 
-      const s = new MagicString(code)
-      let hasReplaced = false
-      let match: RegExpExecArray | null
+export async function replaceDefine(
+  code: string,
+  id: string,
+  define: Record<string, string>,
+  config: ResolvedConfig,
+): Promise<{ code: string; map: string | null }> {
+  // Because esbuild only allows JSON-serializable values, and `import.meta.env`
+  // may contain values with raw identifiers, making it non-JSON-serializable,
+  // we replace it with a temporary marker and then replace it back after to
+  // workaround it. This means that esbuild is unable to optimize the `import.meta.env`
+  // access, but that's a tradeoff for now.
```

Comment:
What is an example of `may contain values with raw identifiers`?
What is the cost of `esbuild is unable to optimize`?

Reply:
And you think this would be to breaking to limit Vite define to the esbuild one?

Reply:
I'm not sure I understand. We would have to continue making our plugins work, and the issue is only specific to `import.meta.env`, so a workaround is needed for `import.meta.env` specifically. The rest user-facing `define` rules still need to follow esbuild's rule.